Hipify revamp

jithunnair-amd opened the pull request:
This PR revamps the hipify module in PyTorch to overcome a long list of shortcomings in the original implementation. However, these improvements are applied only when using hipify to build PyTorch extensions, **not for PyTorch or Caffe2 itself**.

Correspondingly, changes are made to `cpp_extension.py` to match these improvements.

The list of improvements to hipify is as follows:

1. Hipify files in the same directory as the original file, unless there's a "cuda" subdirectory in the original file path, in which case the hipified file will be in the corresponding file path with "hip" subdirectory instead of "cuda".
2. Never hipify the file in-place if changes are introduced due to hipification i.e. always ensure the hipified file either resides in a different folder or has a different filename compared to the original file.
3. Prevent re-hipification of already hipified files. This avoids creation of unnecessary "hip/hip" etc. subdirectories and additional files which have no actual use.
4. Do not write out hipified versions of files if they are identical to the original file. This results in a cleaner output directory, with minimal number of hipified files created.
5. Update header rewrite logic so that it accounts for the previous improvement.
6. Update header rewrite logic so it respects the rules for finding header files depending on whether `""` or `<>` is used.
7. Return a dictionary of mappings of original file paths to hipified file paths from `hipify` function.
8. Introduce a version for hipify module to allow extensions to contain back-compatible code that targets a specific point in PyTorch where the hipify functionality changed.
9. Update `cuda_to_hip_mappings.py` to account for the ROCm component subdirectories inside `/opt/rocm/include`. This also results in cleanup of the `Caffe2_HIP_INCLUDE` path to remove unnecessary additions to the include path.

The list of changes to `cpp_extension.py` is as follows:
1. Call `hipify` when building a CUDAExtension for ROCm.
2. Prune the list of source files to CUDAExtension to include only the hipified versions of any source files in the list (if both original and hipified versions of the source file are in the list)
3. Add subdirectories of /opt/rocm/include to the include path for extensions, so that ROCm headers for subcomponent libraries are found automatically

cc @jeffdaily @sunway513 @hgaspar @lcskrishna @ashishfarmer

jithunnair-amd committed:
Update header hipification logic for ROCm

jithunnair-amd committed:
Missed curand.h

jithunnair-amd committed:
Add all ROCm component include subdirectories in include path when compiling PyTorch extension files

jithunnair-amd committed:
Add version for hipify module

jithunnair-amd committed:
Update naming logic for hipifying PyTorch extensions

jithunnair-amd committed:
Return details about hipified files when calling hipify

jithunnair-amd committed:
Add hipify breadcrumb to C-style files to prevent rehipification

jithunnair-amd committed:
Do not return details of files that were ignored for hipification

jithunnair-amd committed:
Use only relative paths for all_files to ensure no duplicate entries

jithunnair-amd committed:
Hipify as part of building PyTorch cpp extensions for ROCm and replace sources list with hipified sources

jithunnair-amd committed:
Use .append for list to get correct behaviour

jithunnair-amd committed:
Add comments

jithunnair-amd committed:
Use absolute paths in hipify result

jithunnair-amd committed:
Don't make identical hipified copies in same folder, and make hipify recursive to handle header rewrites correctly with that change

jithunnair-amd committed:
Remove extra prints

jithunnair-amd committed:
If include_dirs argument is not provided to CUDAExtension, assume all files are included

jithunnair-amd committed:
Handle header include paths such as '../header.h' in header rewrite logic

jithunnair-amd committed:
Do not write hipify breadcrumb for in-place hipified files

jithunnair-amd committed:
flake8 fixes

t-vi commented:
Cool!

jithunnair-amd committed:
Update test/cpp_extensions/setup.py since hipify is now called within CUDAExtension

jeffdaily commented:
My only concern is that `HIPIFY_FINAL_RESULT` is a module global.  I'm sure it made the implementation easier.  It's probably okay, even if hipify is, for whatever reason, called more than once by the same process.  Right?

jithunnair-amd commented:
> My only concern is that `HIPIFY_FINAL_RESULT` is a module global. I'm sure it made the implementation easier. It's probably okay, even if hipify is, for whatever reason, called more than once by the same process. Right?

Yes, it works fine even if hipify is called multiple times by the same process.

malfet commented:
Just a general comment: when modifying python code, can you please add type annotation decorators (specifically for return types?)

malfet commented:
Also, can you please rebase this PR against `viable/strict` branch to make sure all CI is green?

jithunnair-amd committed:
Add more type annotations

jithunnair-amd commented:
> Just a general comment: when modifying python code, can you please add type annotation decorators (specifically for return types?)

@jeffdaily I added type annotations. Please review https://github.com/pytorch/pytorch/pull/45451/commits/d0e494568bff46ee8b44d1aef6a5cb5d9b1a3b6d

jithunnair-amd committed:
Flake8 fix for line too long

jithunnair-amd committed:
Set is_pytorch_extension=True when calling get_hip_file_path for pytorch extensions

jithunnair-amd commented:
Looks like some connection issues in the `pr/caffe2-pytorch-linux-bionic-rocm3.8-py3.6-test` run. 
```
04:51:05  > git fetch --tags --force --progress https://github.com/pytorch/pytorch.git +refs/heads/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pr/* --depth=70
05:01:05 ERROR: Timeout after 10 minutes
05:01:12 ERROR: Error fetching remote repo 'origin'
05:01:12 hudson.plugins.git.GitException: Failed to fetch from https://github.com/pytorch/pytorch.git
```

@pytorchbot retest this please

ezyang commented:
this seems likely to conflict with #47428, cc @robieta

jithunnair-amd commented:
> this seems likely to conflict with #47428, cc @robieta

@ezyang From a perusal of the changes in that PR, they seem orthogonal to the changes in this PR. Do you think otherwise? If you mean merge conflicts, I can check once that PR is merged.

ezyang commented:
> From a perusal of the changes in that PR, they seem orthogonal to the changes in this PR. Do you think otherwise? If you mean merge conflicts, I can check once that PR is merged.

Yeah, having skimmed this PR they seem mild, probably just merge conflicts.

jithunnair-amd commented:
@pytorchbot retest this please

jithunnair-amd commented:
@ezyang Can someone from the PyTorch team review this PR?

ezyang commented:
Sorry about this delay, wasn't sure if @malfet was gonna grab it. I'll take a look.

malfet commented on torch/utils/cpp_extension.py:
```diff
@@ -821,6 +821,27 @@ def CUDAExtension(name, sources, *args, **kwargs):
 
     include_dirs = kwargs.get('include_dirs', [])
+
+    hipified_sources = set()
```
Nit: move this inside the if block
```suggestion
```

malfet commented on torch/utils/cpp_extension.py:
```diff
@@ -821,6 +821,27 @@ def CUDAExtension(name, sources, *args, **kwargs):
+            is_pytorch_extension=True,
+        )
+
+        for source in sources:
```
```suggestion
        hipified_sources = set()
        for source in sources:
```

malfet committed:
Apply suggestions from code review

Move hipified_sources initialization inside the `if IS_HIP_EXTENSION` block

jithunnair-amd commented:
@malfet Any updates/issues with the PR in the internal builds?

jithunnair-amd committed:
Revert "Add all ROCm component include subdirectories in include path when compiling PyTorch extension files"

This reverts commit 37b80326fcade787bb277ecff17dd417229f9836.

jithunnair-amd commented:
I'm looking into the `test_cpp_extensions_aot_ninja` unit tests failures in the `pr/pytorch-linux-bionic-rocm3.9-py3.6` ROCm build.

jithunnair-amd committed:
Remove ROCm-specific case since CUDA-specific case works for ROCm as well with hipify revamp

jithunnair-amd commented:
Reverted one of the earlier commits in this PR (https://github.com/pytorch/pytorch/pull/45451/commits/2256aa454b664054ee8a9a5218e90087a6d866d7) as it was causing torchvision build failures in CI as the wrong `limits.h` header was being picked up (from `/opt/rocm/include/thrust` instead of `/opt/rocm-3.9.0/llvm/lib/clang/12.0.0/include/`).

Looks like some changes went in while the ROCm CI was broken, which is why we had the `test_cpp_extensions_aot_ninja` unit test failures. I rebased my branch on top of upstream master and added another commit to fix the unit test failures.

jithunnair-amd commented:
@malfet The CI runs passed. I think this PR is ready to be merged now.

jithunnair-amd commented:
@malfet @ezyang Ping

ezyang commented:
Sorry, can you open a new PR for this? Internal system doesn't like it if you reland a previously existing PR

jithunnair-amd commented:
@ezyang Sure, I'll close this and open a new PR. Is it okay to file the new PR from the same branch, or is that an issue for the internal system as well?

ezyang commented:
same branch is OK if github will let you
